### ISSUE TITLE
Urgent - next/image "url" parameter is valid but upstream response is invalid (2)

### ISSUE THREAD

**Issue by bluecomet1994** [vercel/next.js#58248]
### Link to the code that reproduces this issue

https://github.com/bluecomet1994

### To Reproduce

I'm using Next 14.0.1 for the web application development. But I'm facing some issues with the usage of optimized Image component.
The image is not appear on the browser when I run the website and when I go though the network tab, I'm getting this response from them.

next/image "url" parameter is valid but upstream response is invalid

### Current vs. Expected behavior

The output should not have any error response in the network and should display the image with the specific path.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
Windows 10
21H2
19044.1288

Packages
NPM v9.8.1
Yarn v1.22.19
PNPM v8.10.2

node_modules
  "dependencies": {
    "@emotion/react": "^11.11.1",
    "@emotion/styled": "^11.11.0",
    "@mui/icons-material": "^5.14.16",
    "@mui/material": "^5.14.17",
    "next": "^14.0.1",
    "react": "^18",
    "react-dom": "^18",
    "react-redux": "^8.1.3",
    "redux": "^4.2.1",
    "redux-devtools-extension": "^2.13.9",
    "redux-thunk": "^2.4.2"
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "autoprefixer": "^10.0.1",
    "eslint": "^8",
    "eslint-config-next": "14.0.1",
    "postcss": "^8",
    "tailwindcss": "^3.3.0",
    "typescript": "^5"
  }

next.config.js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  i18n: {
    locales: ["en", "es"],
    defaultLocale: "en"
  }
};

module.exports = nextConfig;
```

### Which area(s) are affected? (Select all that apply)

Image optimization (next/image, next/legacy/image)

### Additional context

In earlier versions like below 13.5, it works fine.

**Comment by stasdre** [comment]
Same problem in NextJS v14.0.0

**Comment by cpotey** [comment]
Same issue over here

**Comment by guiluchesi** [comment]
I've been facing the same problem for weeks now :C

**Comment by lostchopstik** [comment]
Experiencing this as well on 14.0.1

**Comment by CromulentCoder** [comment]
I was facing the same issue and spent a whole day debugging this on my end.

In my case, I moved the path of some images from `/some/folder1` to `/public/img`.
After testing the change by running `npm run dev` I was getting the fetch error as above and apparently happened only for .png images. All .svgs worked as intended. 
I verified that the image format was not being blocked by any next.config.js configuration or through any middleware security.

However after running the code with `npm run build` and `npm run start`, I loaded the images and they seemed to work fine. **So for me, the issue was with dev environment**. 
However upon running the dev environment again the images loaded normally. The actual issue was caused in the way next loads the images. The **svgs** were being server from `_next/static` path whereas the the **pngs** were being served with `/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fmyfile.somehash.png"`. Since I moved the folders of the images, the path did not exist in my build folder from which it seems to be generating the hash for referencing the image.


TL;DR Issue in dev environment. Run `npm run build` to add image to build cache and try running dev again.

**Comment by bluecomet1994** [comment]
![Screenshot 2023-11-09 161103](https://github.com/vercel/next.js/assets/134896164/a1058cab-45c3-4973-aaf5-6c802838754b)

But I'm not able to see any image resources after running "npm run build" script.

**Comment by CromulentCoder** [comment]
Try running npm run start and going manually to that page. Maybe its generating the images on the fly?

**Comment by bluecomet1994** [comment]
> Try running npm run start and going manually to that page. Maybe its generating the images on the fly?

Yeah, I did like that, but it is so wired.

**Comment by Naasgul** [comment]
Ive got the same issue with next 14.0.0 - what's weird is that some images on my deployed app are displayed correctly and some get the "url" parameter is valid but upstream response is invalid" error, all png images, all in my public folder

**Comment by stasdre** [comment]
Locally everything is ok for me, but on the firevase hosting the png images don’t work

**Comment by smessaris** [comment]
> Ive got the same issue with next 14.0.0 - what's weird is that some images on my deployed app are displayed correctly and some get the "url" parameter is valid but upstream response is invalid" error, all png images, all in my public folder

I am running into this, or similar as well. Vercel seems to be caching the 502 response if for any reason it couldn't find the source image, and even when the source file is available it keeps serving the 502. Would appreciate a fix.

**Comment by Felixaverlant** [comment]
> Locally everything is ok for me, but on the firevase hosting the png images don’t work

Same here.

**Comment by bluecomet1994** [comment]
> Locally everything is ok for me, but on the firevase hosting the png images don’t work

How do you run it fine on local?

**Comment by Naasgul** [comment]
my webpage loads some pictures correctly in mobile view which arent able to load with known error message on desktop view - I'll test if setting the width/height of the image component helps with solving this issue since its relevant for the optimization (/_next/image?url=%2Fimages%2FHeresMyImage.webp&**w=1080&q=75**)

**Comment by stasdre** [comment]
> How do you run it fine on local?


```javascript
import image from '@/assets/images/image.png';

<Image
   src={image}
   className={'mb-8 h-auto max-w-[110px] md:max-w-[175px] xl:order-last xl:mb-0 xl:shrink-0'}
   width={175}
   height={360}
   quality={85}
   sizes={'100vw'}
   placeholder={'blur'}
/>
``` 

This case locally works& But after deploy to firebase  I get an error: `"url" parameter is valid but upstream response is invalid`

I'm temporarily ditching the component `next/image` for now and just using `<picture>` tag

**Comment by Naasgul** [comment]
> > How do you run it fine on local?
> 
> ```js
> import image from '@/assets/images/image.png';
> 
> <Image
>    src={image}
>    className={'mb-8 h-auto max-w-[110px] md:max-w-[175px] xl:order-last xl:mb-0 xl:shrink-0'}
>    width={175}
>    height={360}
>    quality={85}
>    sizes={'100vw'}
>    placeholder={'blur'}
> />
> ```
> 
> This case locally works& But after deploy to firebase I get an error: `"url" parameter is valid but upstream response is invalid`
> 
> I'm temporarily ditching the component `next/image` for now and just using `<picture>` tag

You can also just use the "unoptimized" attribute for the next/image component - for quick fix I transformed all my .png-images to .webp and use the unoptimized attribute - Im not really seeing a performance issue, having just a small webpage tho

**Comment by stasdre** [comment]
> You can also just use the "unoptimized" attribute for the next/image component - for quick fix I transformed all my .png-images to .webp and use the unoptimized attribute - Im not really seeing a performance issue, having just a small webpage tho

Yes, but this does not solve the problem with the png format. And there is a bug with png images and it needs to be fixed

**Comment by AymanMorsy** [comment]
guys I faced this issue in **v14.0.1** but resolved with **v14.0.2** give it a try and see

**Comment by AymanMorsy** [comment]
**I'm confirming this bug in v14.0.2 and it is so weird bug sometimes display and sometimes no issue, refreshing the page can reproduce it. somtimes close the browser and deleting the build directory then running `npm run dev` it sometimes reproduce it**

```
windows 10 pro v22H2 (19045.3570)
node v18.15.0
 "dependencies": {
    "next": "^14.0.2",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^18.2.34",
    "@types/react-dom": "^18.2.14",
    "autoprefixer": "^10",
    "eslint": "^8",
    "eslint-config-next": "^14.0.1",
    "postcss": "^8",
    "tailwindcss": "^3",
    "typescript": "^5"
 ```

**Comment by AymanMorsy** [comment]
**I'm confirming this bug in v14.0.2 and it is so weird bug sometimes display and sometimes no issue, refreshing the page can reproduce it. somtimes close the browser and deleting the build directory then running `npm run dev` it sometimes reproduce it**

```
windows 10 pro v22H2 (19045.3570)
node v18.15.0
 "dependencies": {
    "next": "^14.0.2",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^18.2.34",
    "@types/react-dom": "^18.2.14",
    "autoprefixer": "^10",
    "eslint": "^8",
    "eslint-config-next": "^14.0.1",
    "postcss": "^8",
    "tailwindcss": "^3",
    "typescript": "^5"
 ```

**Comment by ihnokim** [comment]
Same here with v14.0.0 :( ... any improvements?

**Comment by joey-ma** [comment]
will someone try `npm ci` and see if it resolves the issue?

https://docs.npmjs.com/cli/v8/commands/npm-ci

**Comment by jmacbhc** [comment]
Same here for IIS.

**Comment by evgenTraytyak** [comment]
for me it's only for IIS (works well on Heroku/Vercel)
v.13.5.5

**Comment by ErosKarm** [comment]
Issue still apparant.

**Comment by a-rbsn** [comment]
I'm in `14.0.3` and I'm trying to load an image from CraftCMS in my local dev environment using GraphQL, the query pulls the image url through fine but `next/image` won't load it, I get this on console:

```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:14152:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async imageOptimizer (/Users/andrewrobinson/Sites/unravelyourfood/node_modules/next/dist/server/image-optimizer.js:560:29)
    at async cacheEntry.imageResponseCache.get.incrementalCache (/Users/andrewrobinson/Sites/unravelyourfood/node_modules/next/dist/server/next-server.js:176:65)
    at async /Users/andrewrobinson/Sites/unravelyourfood/node_modules/next/dist/server/response-cache/index.js:90:36
    at async /Users/andrewrobinson/Sites/unravelyourfood/node_modules/next/dist/lib/batcher.js:41:32 {
  cause: Error: unable to verify the first certificate
      at TLSSocket.onConnectSecure (node:_tls_wrap:1540:34)
      at TLSSocket.emit (node:events:513:28)
      at TLSSocket._finishInit (node:_tls_wrap:959:8)
      at ssl.onhandshakedone (node:_tls_wrap:743:12) {
    code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE'
  }
}
```

**Comment by a-rbsn** [comment]
I'm in `14.0.3` and I'm trying to load an image from CraftCMS in my local dev environment using GraphQL, the query pulls the image url through fine but `next/image` won't load it, I get this on console:

```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:14152:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async imageOptimizer (/Users/andrew/Sites/unravelyourfood/node_modules/next/dist/server/image-optimizer.js:560:29)
    at async cacheEntry.imageResponseCache.get.incrementalCache (/Users/andrew/Sites/unravelyourfood/node_modules/next/dist/server/next-server.js:176:65)
    at async /Users/andrew/Sites/unravelyourfood/node_modules/next/dist/server/response-cache/index.js:90:36
    at async /Users/andrew/Sites/unravelyourfood/node_modules/next/dist/lib/batcher.js:41:32 {
  cause: Error: unable to verify the first certificate
      at TLSSocket.onConnectSecure (node:_tls_wrap:1540:34)
      at TLSSocket.emit (node:events:513:28)
      at TLSSocket._finishInit (node:_tls_wrap:959:8)
      at ssl.onhandshakedone (node:_tls_wrap:743:12) {
    code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE'
  }
}
```

**Comment by Muffeter** [comment]
@bluecomet1994 I think this probably not the bug for next.

### TL;DR 
Just explicitly specify the port like this ( in this example [luca-vogels/bug-next-image](https://github.com/luca-vogels/bug-next-image)) : 
```
const dev = true;
const HTTP_PORT = 80;
const nextApp = next({ dev, port: HTTP_PORT });
```
And it'll run perfectly without any error.

![image](https://github.com/vercel/next.js/assets/36598604/48481aa4-0b7a-449d-a6e2-96cf83926df9)

I also test latest next.js version `next@14.0.4-canary.5`, and everything is good.


### WHY

The Component <Image /> always optimize the image. 
```
unoptimized = {false} // {false} | {true}
```
And when client want the image, server will call [imageOptimizer](https://github.com/vercel/next.js/blob/ce183f6e3e3c532b0fb1343914e5369cee7f7920/packages/next/src/server/next-server.ts#L521C22-L521C22) to optimize the image. So you can specify unoptimized in <Image /> tag to solve this problem (that's why @Naasgul 's method work), but it not the best choice.

And [In this line](https://github.com/vercel/next.js/blob/ce183f6e3e3c532b0fb1343914e5369cee7f7920/packages/next/src/server/next-server.ts#L551C6-L551C6) next server try to fetch image with url `http://localhost:<port>`
Here is the bug:
```
const invokeRes = await invokeRequest(
           `${protocol}://${this.fetchHostname || 'localhost'}:${this.port}${newReq.url || ''}`,
            {
              method: newReq.method || 'GET',
              headers: newReq.headers,
              signal: signalFromNodeResponse(res.originalResponse),
			}
)
```

If you don't pass the port explicitly like `next({ dev, port: HTTP_PORT })`, next don't know the port, and `this.port` will be `3000`, which is the default port of next.

**Comment by jacksonllk** [comment]
I have just updated to Next 14.0.3 and facing the same issue. My log on render.com:
![Screenshot 2023-11-22 151310](https://github.com/vercel/next.js/assets/96425149/5d45e928-3298-4f68-954c-7bdb4b028a26)

**Comment by stx-chris** [comment]
Just recently found the reason why Next.js was giving us this error. It was due to Next trying to convert our local `.webp` images into `.avif` format for Mac browsers.

When this line is used in `next.config.js`, the issue appears and in addition causes a timeout after 40 seconds or so:

```
    formats: ["image/avif", "image/webp"],
```

Once we removed the `formats` property entirely, the issue went away. Might have to do with the `sharp` package integration.

**Comment by jacksonllk** [comment]
> Just recently found the reason why Next.js was giving us this error. It was due to Next trying to convert our local `.webp` images into `.avif` format for Mac browsers.
> 
> When this line is used in `next.config.js`, the issue appears and in addition causes a timeout after 40 seconds or so:
> 
> ```
>     formats: ["image/avif", "image/webp"],
> ```
> 
> Once we removed the `formats` property entirely, the issue went away. Might have to do with the `sharp` package integration.

But i currently do not have the format property in my next.config.js and I am still seeing that error

**Comment by stx-chris** [comment]
Do you import local files as recommended in the [docs](https://nextjs.org/docs/pages/building-your-application/optimizing/images#local-images) using `import` statements or do you use a relative path to some folder in the `src` property? 

Because switching from relative path to imports helped us reduce potential errors with image loading.

**Comment by jacksonllk** [comment]
I tried doing that but it doesn't work for me somehow. My image component is like so: 
      <Image
        src="/images/hero-image1.webp"
        alt="Hero Image"
        fill={true}
        style={{ 
          objectFit: 'cover',
          objectPosition: 'top',
          }} 
        quality={10}
        priority 
      />

I'm still on pages and not app router. Not sure if that is causing it. So the error in this thread is related to images not being imported?

**Comment by stx-chris** [comment]
ok, can you try the following instead?

```
// the relative path might need adjustment according to your file location
import heroImage1 from "/images/hero-image1.webp"; 

<Image
   src={heroImage1}
   alt="Hero Image"
   fill={true}
   style={{
       objectFit: 'cover',
       objectPosition: 'top',
   }}
   quality={10}
   priority
/>
```

Not sure in which version this way of handling was introduced (whether only for app router or already for pages), just give it a try. Good luck!

**Comment by jacksonllk** [comment]
Yea for some reason this doesn't work for me, but thanks for your help! Still waiting for a definitive answer on the error of this thread. It's been unresolved since August.

**Comment by hsnbsst** [comment]
same problem exists

**Comment by wlemahieu** [comment]
The issue seems be related to the Image Optimizer expecting to find the images on port 3000 in production. When I checked my production logs, I see the error along the lines of "econnrefused 127.0.0.1 PORT 3000".

I run my project locally on port 4000, but in production it's using port 8080 to serve the web app, not 3000... I can see the image fine in production when I visit the exact URL of the image (served from /public/*) in my browser!

The way I'm getting around this is to use an absolute URL of the image for the `Image` `src`. I then use an `ENV` variable that I have defined both on my production server as well as in my env.local file.

**Workaround Example**

_Production:_
`APPURL=https://some-remote-server`

_Local:_
`APPURL=http://localhost:4000`

Then in my component use string interpolation:

``<Image src={`${process.env.APPURL}/image.png`} />``

**Comment by CemYil03** [comment]
We are currently experienceing the same issue with version `14.0.3`. May it be related to the way the app is hosted? We run it in plesks passenger, which does an automatic foreward to the active port of the nodejs application.

**Comment by E1NSER** [comment]
> We are currently experienceing the same issue with version `14.0.3`. May it be related to the way the app is hosted? We run it in plesks passenger, which does an automatic foreward to the active port of the nodejs application.

I had the same issue with Plesk Passenger. The only way to get it running was to make the paths to the images absolute.

**Comment by sleeporsmile** [comment]
same issue. self hosting in docker container

**Comment by PawelWywiol** [comment]
at the moment this works for me https://github.com/vercel/next.js/issues/58248#issuecomment-1820934773

I tried it before, but with different port then `80`

I'm using hosting with Plex Passanger

**Comment by virajstrivedge** [comment]
> at the moment this works for me [#58248 (comment)](https://github.com/vercel/next.js/issues/58248#issuecomment-1820934773)
> 
> I tried it before, but with different port then `80`
> 
> I'm using hosting with Plex Passanger

I am also using plesk i can you explain step for setup currently i am using server.js file but image issue there

**Comment by shino369** [comment]
issue still persist in 14.0.3

**Comment by PhAlves23** [comment]
We are on version 14.0.1 and we have configured the loader and solved our problem

`<Image src={profilePhoto ? profilePhoto : '/user.png'} 
width={170} 
height={170} 
loader={({ src, width, quality }: ImageLoaderProps) => { return `${ profilePhoto ? profilePhoto : '/user.png' }?w=${width}&q=${quality || 75}` }} 
alt="Person Image" 
className="mb-3 max-h-[170px] min-h-[170px] min-w-[170px] max-w-[170px] rounded-full object-cover" />`

**Comment by corazon-roto** [comment]
This worked for me. I guess the problem is related to the Http and https itself. I know `domains` is deprecated, just testing around.
```

  images: {
    domains: ['localhost', process.env.NEXT_PUBLIC_SERVER_URL]
      .filter(Boolean)
      .map(url => url.replace(/https?:\/\//, '')),
  },

```

**Comment by corazon-roto** [comment]
This worked for me. I guess the problem is related to the Http and https itself. I know `domains` is deprecated, just testing around.

```
images: {
    domains: ['localhost', process.env.NEXT_PUBLIC_SERVER_URL]
      .filter(Boolean)
      .map(url => url.replace(/https?:\/\//, '')),
  },

```

**Comment by bluecomet1994** [comment]
> We are on version 14.0.1 and we have configured the loader and solved our problem
> 
> ```js
> <Image
>         src={profilePhoto ? profilePhoto : '/user.png'}
>         width={72}
>         height={72}
>         loader={({ src, width, quality }: ImageLoaderProps) => {
>           return `${profilePhoto ? profilePhoto : '/user.png'}?w=${width}&q=${
>             quality || 75
>           }`
>         }}
>           alt="Person Image"
>           className="max-h-[72px] min-h-[72px] min-w-[72px] max-w-[72px] cursor-pointer rounded-full object-cover"
>         />
> ```

Should I have to apply that attribute to all Image components?

**Comment by lukasnevosad** [comment]
After 2 days of fiddling with different Next.js versions to no effect, I realized this only affects local images, not images loaded from other websites.

So a **simple workaround** is to use absolute URLs in `<Image>` `src` attribute. You then also need to enable the domain in `nextConfig.images.remotePatterns` in `next.config.js`.

**Comment by lukasnevosad** [comment]
> @lukasnevosad, but this does also mean that you don't make use of the image optimization of Next.js anymore, right?

It will still optimize the image, that's why you need to enable the domain in next.config.js

**Comment by mo-dhia** [comment]
this can also be vercel image optimization limits, read more here
https://vercel.com/docs/image-optimization/managing-image-optimization-costs

**Comment by shkatulo** [comment]
> The issue seems be related to the Image Optimizer expecting to find the images on port 3000 in production. When I checked my production logs, I see the error along the lines of "econnrefused 127.0.0.1 PORT 3000".
> 
> I run my project locally on port 4000, but in production it's using port 8080 to serve the web app, not 3000... I can see the image fine in production when I visit the exact URL of the image (served from /public/*) in my browser!
> 
> The way I'm getting around this is to use an absolute URL of the image for the `Image` `src`. I then use an `ENV` variable that I have defined both on my production server as well as in my env.local file.
> 
> **Workaround Example**
> 
> _Production:_ `APPURL=https://some-remote-server`
> 
> _Local:_ `APPURL=http://localhost:4000`
> 
> Then in my component use string interpolation:
> 
> `` <Image src={`${process.env.APPURL}/image.png`} /> ``

This solution worked for me. But I also had to add `NODE_TLS_REJECT_UNAUTHORIZED=0` to my `dev` script in package.json to make it working locally with self-signed certificate.

**Comment by Michle99** [comment]
Had the same error in next@14.0.4 with using the `.jpg` & `.png` on the client side.

```
import Image from "next/image";
import logo from "../../../public/icons/logo.jpg";

<Image 
  src={logo} 
  blurDataURL="/app/icons/logo.jpg"
  width={25} 
  height={25} 
  alt="logo" 
/>
```
Error:
`
⨯ upstream image response failed for /_next/static/media/logo.8a5e6fc1.jpg TypeError: fetch failed
`

node v18.17.0

```
"dependencies": {
    "next": "14.0.4",
    "next-sanity": "^7.0.4",
    "react": "^18",
    "react-dom": "^18",
    "sanity": "^3.21.3",
    "styled-components": "^6.1.1"
  },
  "devDependencies": {
    "@types/node": "^20",
    "@types/react": "^18",
    "@types/react-dom": "^18",
    "autoprefixer": "^10.0.1",
    "eslint": "^8",
    "eslint-config-next": "14.0.4",
    "postcss": "^8",
    "react-icons": "^4.12.0",
    "tailwindcss": "^3.3.0",
    "typescript": "^5"
  }
```

**Comment by 6feferonka9** [comment]
Another one with an issue. Im running a custom server.js on DEV with TSL, even setting the port to 443 on
const app = next({ dev, port }); gives an error: 

```
    code: 'UND_ERR_SOCKET',
    socket: {
      localAddress: '127.0.0.1',
      localPort: 60476,
      remoteAddress: '127.0.0.1',
      remotePort: 443,
      remoteFamily: 'IPv4',
      timeout: undefined,
      bytesWritten: 946,
      bytesRead: 0
    }
```

**Comment by vesters** [comment]
> > We are currently experienceing the same issue with version `14.0.3`. May it be related to the way the app is hosted? We run it in plesks passenger, which does an automatic foreward to the active port of the nodejs application.
> 
> I had the same issue with Plesk Passenger. The only way to get it running was to make the paths to the images absolute.

Does this mean that you miss out on the image optimization? I have the same issue with plesk and my next app :-(

**Comment by vesters** [comment]
Btw. using next 14.0.4 and sharp 0.32.6.

**Comment by samcx** [comment]
@bluecomet1994 Please add a complete <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture> so we can take a look at this! At the moment, you are just linking your <picture data-single-emoji=":github:" title=":github:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/github/16d8edd1631df957.png" alt=":github:" align="absmiddle"></picture> profile. I added the label as a explaining for creating a <picture data-single-emoji=":repro:" title=":repro:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/repro/5cea53b5532825b4.png" alt=":repro:" align="absmiddle"></picture>.

**Comment by LuLue7775** [comment]
> , but thanks for your help! Still waitin

Can we possibly fix this by explicitly designate port setting on next.config?

**Comment by iuliancmarcu** [comment]
@samcx I set up this repository that has the minimal setup for reproducing this issue:

https://github.com/iuliancmarcu/nextjs-local-image-issue-repro

**Instructions:**
1. clone project
2. run `create_cert_and_add_hostname.sh` - this will self-sign a certificate, add it to the keychain, and add a `local.mycorp.app` entry in `/etc/hosts`
3. run `npm install && node server.js`
4. open `https://local.mycorp.app:3000` in your browser and see the image not loading

**Comment by qcasey** [comment]
Separate from the TLS issues some are facing, I ran into this error when importing and configuring [shipengine-js](https://github.com/ShipEngine/shipengine-js). Was at my wits end and just started cutting files from my `app/page.tsx`.

I removed the package from my codebase and haven't gotten the `upstream response is invalid` error since, on either dev or prod, and local images load at cold start.

I have no idea why (esm/cjs or node implementation?), but maybe check your deps as a workaround.

**Comment by SEI-John** [comment]
I also experienced this issue with 14.0.4 and the quick fix for me was to use the loader attribute from @PhAlves23

> We are on version 14.0.1 and we have configured the loader and solved our problem
> 
> ```js
> <Image
>         src={profilePhoto ? profilePhoto : '/user.png'}
>         width={72}
>         height={72}
>         loader={({ src, width, quality }: ImageLoaderProps) => {
>           return `${profilePhoto ? profilePhoto : '/user.png'}?w=${width}&q=${
>             quality || 75
>           }`
>         }}
>           alt="Person Image"
>           className="max-h-[72px] min-h-[72px] min-w-[72px] max-w-[72px] cursor-pointer rounded-full object-cover"
>         />
> ```

**Comment by Innei** [comment]
I also encountered this problem. My CDN image service prohibits access to requests that do not carry the HTTP header Reference. Now, Next's image optimization does not forward the original request's HTTP header, which causes my images to be inaccessible.

**Comment by 22chrs** [comment]
Same here with balena cloud. I have a multi docker environment in which I have one container running a chromium browser, and one running my next js app. After upgrading next js optimized images now longer load.

**Comment by apavlinovic** [comment]
I am also encountering this problem:

- 14.1.0, the app runs on port 3000, AVIF/WEBP enabled, self hosted
- Prior to today we were running 13.5.2 with no issues, the only difference is the 13.5.2 to 14.1.0 upgrade of Next JS

**Comment by styfle** [comment]
@iuliancmarcu Thank you for the [steps to reproduce the issue](https://github.com/vercel/next.js/issues/58248#issuecomment-1884901487) 🙌 

I've narrowed it down to 13.4.12 as the last known good version and 13.4.13 is when it broke.

I believe that previously, "external" images were fetched but "internal" images were not. However, the image optimization code now runs in a different process so "internal" images needed to be fetched now. See this line:

https://github.com/vercel/next.js/blob/16dcebe79270985782307c19a97ab1755285e3c8/packages/next/src/server/lib/server-ipc/invoke-request.ts#L22

When running https with a custom server, its failing in this case because the `NODE_EXTRA_CA_CERTS` env is no longer preserved and the protocol is not preserved. See protocol selection here:

https://github.com/vercel/next.js/blob/16dcebe79270985782307c19a97ab1755285e3c8/packages/next/src/server/next-server.ts#L563-L570

If we want to continue to use IPC, then we'll probably need custom servers to specify they are using https because hostname and port is no longer enough. See custom server here: https://github.com/iuliancmarcu/nextjs-local-image-issue-repro/blob/433aca7587482dd78e8cf8f4322e5665c5707084/server.js#L11C13-L11C17

I'll discuss possible solutions with the @vercel/next-js team this week so we can figure out the best path forward, thanks!

**Comment by aplr** [comment]
When self-hosting nextjs, and setting `HOSTNAME=0.0.0.0` the fetch request will also fail, however with another error.

This is, as the fetch hostname is set to the specified hostname here:

https://github.com/vercel/next.js/blob/16dcebe79270985782307c19a97ab1755285e3c8/packages/next/src/server/base-server.ts#L437-L440

This is the error I'm getting

```
 ⨯ upstream image response failed for /_next/static/media/PhoneMock.ef782fce.png TypeError: fetch failed
    at node:internal/deps/undici/undici:12344:11
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async invokeRequest (/app/node_modules/.pnpm/next@14.0.4_@babel+core@7.22.10_@opentelemetry+api@1.7.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/lib/server-ipc/invoke-request.js:17:12)
    at async /app/node_modules/.pnpm/next@14.0.4_@babel+core@7.22.10_@opentelemetry+api@1.7.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/next-server.js:596:35
    at async imageOptimizer (/app/node_modules/.pnpm/next@14.0.4_@babel+core@7.22.10_@opentelemetry+api@1.7.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/image-optimizer.js:576:13)
    at async cacheEntry.imageResponseCache.get.incrementalCache (/app/node_modules/.pnpm/next@14.0.4_@babel+core@7.22.10_@opentelemetry+api@1.7.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/next-server.js:176:65)
    at async /app/node_modules/.pnpm/next@14.0.4_@babel+core@7.22.10_@opentelemetry+api@1.7.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/response-cache/index.js:90:36
    at async /app/node_modules/.pnpm/next@14.0.4_@babel+core@7.22.10_@opentelemetry+api@1.7.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/lib/batcher.js:41:32 {
  cause: Error: connect ECONNREFUSED 0.0.0.0:3000
      at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1595:16)
      at TCPConnectWrap.callbackTrampoline (node:internal/async_hooks:130:17) {
    errno: -111,
    code: 'ECONNREFUSED',
    syscall: 'connect',
    address: '0.0.0.0',
    port: 3000
  }
}
```

Guess certain systems don't resolve `0.0.0.0` to `127.0.0.1` per default. Google Cloud Run, at least, does not, as far as I can tell.

**Comment by 22chrs** [comment]
going back to 13.4.12 does not solve the problem for me. still no nextjs optimized images. other images are not affected:2024-01-29T01:01:14+01:00  app      errno: -111,
2024-01-29T01:01:14+01:00  app      code: 'ECONNREFUSED',
2024-01-29T01:01:14+01:00  app      syscall: 'connect',
2024-01-29T01:01:14+01:00  app      address: '127.0.0.1',
2024-01-29T01:01:14+01:00  app      port: 3000
2024-01-29T01:01:14+01:00  app    }
2024-01-29T01:01:14+01:00  app  }

**Comment by DedaDev** [comment]
> going back to 13.4.12 does not solve the problem for me. still no nextjs optimized images. other images are not affected:
> 
> 2024-01-29T01:01:14+01:00 app errno: -111, 2024-01-29T01:01:14+01:00 app code: 'ECONNREFUSED', 2024-01-29T01:01:14+01:00 app syscall: 'connect', 2024-01-29T01:01:14+01:00 app address: '127.0.0.1', 2024-01-29T01:01:14+01:00 app port: 3000 2024-01-29T01:01:14+01:00 app } 2024-01-29T01:01:14+01:00 app }

It solves the "upstream" issue for me, but images are not optimized.

**Comment by 89224747505** [comment]
<Image
        src={profilePhoto ? profilePhoto : '/user.png'}
        width={72}
        height={72}
        loader={({ src, width, quality }: ImageLoaderProps) => {
          return `${profilePhoto ? profilePhoto : '/user.png'}?w=${width}&q=${
            quality || 75
          }`
        }}
          alt="Person Image"
          className="max-h-[72px] min-h-[72px] min-w-[72px] max-w-[72px] cursor-pointer rounded-full object-cover"
        />

It's work for me

**Comment by dima-bendik** [comment]
@styfle any updates on this one?
